### PR TITLE
Update blueprint export design

### LINKED ISSUE #53675: Blueprint: Export - Apply the latest design
Figma: [Y5pUYSJPsGEud1vknUZhi8-fi-3814_18301](https://href.li/?https://www.figma.com/file/Y5pUYSJPsGEud1vknUZhi8/?node-id=3814-18301)

To ensure more user friendly U.I, apply the latest design from Figma.

## Acceptance criteria

- WooCommerce -> Settings -> Advanced -> Blueprint should show the latest design
- Users are able to select/de-select settings, plugins, and themes

### PR DESCRIPTION
**Pull request by moon0326**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

Update the current Blueprint export screen to the latest design.

Figma: Y5pUYSJPsGEud1vknUZhi8-fi-3814_18301

### Changes proposed in this Pull Request:

This PR applies the latest design for the "Export" section.

Closes #53675

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with this branch and `blueprint` feature enabled.
2. Navigate to `WooCommerce -> Settings -> Advanced -> Blueprint`
3. Confirm the new design changes for the `Export` section. 
4. Select/De-select some items and click on `Export` button.
5. Inspect the schema and confirm your selections are included.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

This is under a feature-flag.
</details>

### CODE REVIEW

**Review by chihsuan — COMMENTED** [review]
Hey @moon0326 I followed the instructions in the PR description but don't see the new export section in the Blueprint screen.  The `window.wcSettings?.admin?.blueprint_step_groups` is undefined. Is there anything I need to do to make it work? 

Screenshot 2024-12-12 at 11 59 00


> This PR applies the latest design for the "Export" section.


This PR also introduces some changes like PubSub on PHP side. Could you explain more about the changes? It would help me know what to test. Thanks! 🙏 

Besides, I noticed there are some linting errors in wc admin and blueprint packages. It would be great to fix them.

**Comment by moon0326** [comment]
> Hey @moon0326 I followed the instructions in the PR description but don't see the new export section in the Blueprint screen. The `window.wcSettings?.admin?.blueprint_step_groups` is undefined. Is there anything I need to do to make it work?
> 
> Screenshot 2024-12-12 at 11 59 00
> 
> > This PR applies the latest design for the "Export" section.
> 
> This PR also introduces some changes like PubSub on PHP side. Could you explain more about the changes? It would help me know what to test. Thanks! 🙏
> 
> Besides, I noticed there are some linting errors in wc admin and blueprint packages. It would be great to fix them.

My bad. I think there is an issue with screen I.D. The JS var should be registered on the settings page. I'll look into it.

**Comment by moon0326** [comment]
> This PR also introduces some changes like PubSub on PHP side. Could you explain more about the changes? It would help me know what to test. Thanks! 🙏
> 
> Besides, I noticed there are some linting errors in wc admin and blueprint packages. It would be great to fix them.

@chihsuan Could you please try again 🙏 ? The var should populate correctly now.

> This PR also introduces some changes like PubSub on PHP side. Could you explain more about the changes? It would help me know what to test. Thanks! 🙏

It's just [pub-sub pattern](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) implementation. It's very similar to do_action/apply_action, but limited to the object itself (not global). For this PR, it's used in https://github.com/woocommerce/woocommerce/pull/53648/files#diff-07885c25700fecec6f966f25f834bc75a29f4159aeae0bfad270c37eef937c3eR120 to filter the plugin list. We can test it by selecting/deselecting plugin list from the setting page.

> Besides, I noticed there are some linting errors in wc admin and blueprint packages. It would be great to fix them.

Fixed!

File: plugins/woocommerce/src/Admin/Features/Blueprint/Init.php [hunk 1]
```diff
@@ -112,4 +116,90 @@ public function add_woo_importers( array $importers ) {
...
+			return $settings;
+		}
+
+		$screen_id = PageController::get_instance()->get_current_screen_id();
+		if ( 'woocommerce_page_wc-admin' === $screen_id ) {
+			// Add upload nonce to global JS settings. The value can be accessed at wcSettings.admin.blueprint_upload_nonce.
+			$settings['blueprint_upload_nonce'] = wp_create_nonce( 'blueprint_upload_nonce' );
+		}
+
+		wplog( 'screen_id: ' . $screen_id );
```
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/Features/Blueprint/Init.php]: I think we should remove this? I'm getting a fatal error:

```
13-Dec-2024 01:49:34 UTC] PHP Fatal error:  Uncaught Error: Call to undefined function Automattic\WooCommerce\Admin\Features\Blueprint\wplog() in /srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.6.0-dev-12302889822-gedc19a8/src/Admin/Features/Blueprint/Init.php:195
Stack trace:
#0 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Admin\Features\Blueprint\Init->add_js_vars(Array)
#1 /srv/htdocs/__wp__/wp-includes/plugin.php(205): WP_Hook->apply_filters(Array, Array)
#2 /srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.6.0-dev-12302889822-gedc19a8/src/Internal/Admin/WCAdminSharedSettings.php(66): apply_filters('woocommerce_adm...', Array)
#3 /srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.6.0-dev-12302889822-gedc19a8/src/Blocks/Assets/AssetDataRegistry.php(272): Automattic\WooCommerce\Internal\Admin\WCAdminSharedSettings->Automattic\WooCommerce\Internal\Admin\{closure}()
#4 /srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.6.0-dev-12302889822-gedc19a8/src/Blocks/Assets/AssetDataRegistry.php(388): Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry->execute_lazy_data()
#5 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(324): Automattic\WooCommerce\Blocks\Assets\AssetDataRegistry->enqueue_asset_data('')
#6 /srv/htdocs/__wp__/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters('', Array)
#7 /srv/htdocs/__wp__/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#8 /srv/htdocs/__wp__/wp-admin/admin-footer.php(95): do_action('admin_print_foo...')
#9 /srv/htdocs/__wp__/wp-admin/admin.php(297): require_once('/srv/htdocs/__w...')
#10 {main}
  thrown in /srv/htdocs/wp-content/plugins/wc_beta_tester_live_branch_9.6.0-dev-12302889822-gedc19a8/src/Admin/Features/Blueprint/Init.php on line 195
```

File: packages/php/blueprint/src/ExportSchema.php [hunk 2]
```diff
@@ -101,4 +102,12 @@ function ( $exporter ) {
 
 		return $schema;
 	}
+
+	public function onBeforeExport($step_name, $callback) {
```
Comment (chihsuan) [inline packages/php/blueprint/src/ExportSchema.php:106]: Should we also fix lint issues for code in the blueprint package?

**Review by chihsuan — COMMENTED** [review]
Thanks for the fix and explanation! @moon0326 

However, it still doesn't work for me. After removing the `wp_log` code, I can see the full Export section but when I click on the `Export` button, it doesn't trigger the export process. It returns a 500 error.


https://github.com/user-attachments/assets/a1a4b1c4-1ddf-410a-923e-234c41725da7

I think it might be worth adding an error-handling/message to let users know that the export process is not working in a follow-up PR. What do you think?

I also found some UI issues:


1. `Export Blueprint` description width seems to be too short. It should be 598px.
Screenshot 2024-12-13 at 09 57 02



2. The spacing between Toggle button and option should be 16px.

Screenshot 2024-12-13 at 10 02 05


3. The spacing between each option should be 12px for Settings section and 16px for other sections.

Screenshot 2024-12-13 at 10 02 16

4. Section and option description line height should be 20px.
5. Since we don't include all items in `WooCommerce | Settings`, I think we should change the description "It includes all the items featured in WooCommerce | Settings.".

File: plugins/woocommerce/src/Admin/Features/Blueprint/Init.php (hunk 1 above)
Comment (moon0326) [reply plugins/woocommerce/src/Admin/Features/Blueprint/Init.php]: My apologies 🤦 Used it to debug the JS var issue and forgot to remove.

File: packages/php/blueprint/src/ExportSchema.php (hunk 2 above)
Comment (moon0326) [reply packages/php/blueprint/src/ExportSchema.php:106]: Lint isn't setup for the package yet. I'll create a new issue for it.

**Comment by moon0326** [comment]
> Thanks for the fix and explanation! @moon0326
> 
> However, it still doesn't work for me. After removing the `wp_log` code, I can see the full Export section but when I click on the `Export` button, it doesn't trigger the export process. It returns a 500 error.
> 
>  Screen.Recording.2024-12-13.at.10.05.38.mov 
> I think it might be worth adding an error-handling/message to let users know that the export process is not working in a follow-up PR. What do you think?
> 
> I also found some UI issues:
> 
> 1. `Export Blueprint` description width seems to be too short. It should be 598px.
>    Screenshot 2024-12-13 at 09 57 02
> 2. The spacing between Toggle button and option should be 16px.
> 
> Screenshot 2024-12-13 at 10 02 05
> 
> 3. The spacing between each option should be 12px for Settings section and 16px for other sections.
> 
> Screenshot 2024-12-13 at 10 02 16
> 
> 4. Section and option description line height should be 20px.
> 5. Since we don't include all items in `WooCommerce | Settings`, I think we should change the description "It includes all the items featured in WooCommerce | Settings.".

Thanks @chihsuan. 500 error must be coming from composer not installing the latest changes. It works for me locally. I'll address them soon.

**Comment by moon0326** [comment]
> However, it still doesn't work for me. After removing the `wp_log` code, I can see the full Export section but when I click on the `Export` button, it doesn't trigger the export process. It returns a 500 error.

Please try again when `Build Live branch` job is done. There was an error with trait import.

> I think it might be worth adding an error-handling/message to let users know that the export process is not working in a follow-up PR. What do you think?

Good idea. It handles errors if they are known errors, but not 500. We should take care of 500 errors as well.

File: plugins/woocommerce/src/Admin/Features/Blueprint/Init.php (hunk 1 above)
Comment (chihsuan) [inline plugins/woocommerce/src/Admin/Features/Blueprint/Init.php]:
```suggestion
```

**Comment by chihsuan** [comment]
> Please try again when Build Live branch job is done. There was an error with trait import.

Thanks, confirmed export works now!

**Comment by moon0326** [comment]
> Thanks, confirmed export works now!

Thank you!

I've fixed CSS issues in [CSS fixes](https://github.com/woocommerce/woocommerce/pull/53648/commits/8f2fd2bf2d35663581d1a1f5f4c6882619c9bdc0)

For the copies, I think we can review all the copies later. As far as I understand, most of them are not final yet

**Review by chihsuan — APPROVED** [review]
Thanks for addressing the CSS issues! @moon0326 

Just remaining one minor issue:

- The gap between each item in `Plugins and extensions` and `Themes` should be 16px, which is different from the `Settings` section items.
Screenshot 2024-12-17 at 12 39 46

Other than that, looks good!

Pre-approving! 👍

**Pull request merged**